rawfile/rawfile_servicer.py

```diff
@@ -254,29 +205,49 @@ def CreateVolume(self, request: csi_pb2.CreateVolumeRequest, context):
             context.abort(
                 grpc.StatusCode.INVALID_ARGUMENT, "Topology key not found... why?"
             )
-        try:
-            init_rawfile(
-                volume_id=request.name,
-                size=size,
-                thin_provision=thin_provision,
-                freezefs=freezefs,
-                copy_on_write=copy_on_write,
-                snapshot_id=snapshot_id,
-                temporary_snapshot=remove_source_snapshot,
-            )
-        except CalledProcessError as exc:
-            if exc.returncode == RESOURCE_EXHAUSTED_EXIT_CODE:
-                context.abort(
-                    grpc.StatusCode.RESOURCE_EXHAUSTED, "Not enough disk space"
+        if request.volume_content_source:
+            if all(
+                (
+                    request.volume_content_source.snapshot,
+                    request.volume_content_source.snapshot.snapshot_id,
                 )
-            else:
-                raise exc
-        finally:
-            if remove_source_snapshot and snapshot_id:
-                vol_id, snap_name = snapshot_id.rsplit("/", 1)
-                snapshot_manager.delete_snapshot(
-                    volume_id=vol_id, name=snap_name, temporary=remove_source_snapshot
+            ):
+                source_type = VolumeSource.snapshot
+                source_id = request.volume_content_source.snapshot.snapshot_id
+            elif all(
+                (
+                    request.volume_content_source.volume,
+                    request.volume_content_source.volume.volume_id,
                 )
+            ):
+                source_type = VolumeSource.volume
+                source_id = request.volume_content_source.volume.volume_id
+        if utils.storage_pool.get_capacity() < size:
+            context.abort(grpc.StatusCode.RESOURCE_EXHAUSTED, "Not enough disk space")
+        is_ready = False
+        try:
+            is_ready = metadata(request.name).get("ready", False)
+        except FileNotFoundError:
+            self._task_manager.run_task(
+                TaskName.CREATE_VOLUME,
+                request.name,
+                size,
+                thin_provision,
+                freezefs,
+                copy_on_write,
+                source_type,
+                source_id,
+            )
+        start_time = time.time()
+        while time.time() - start_time < 30:
+            is_ready = metadata_or(request.name).get("ready", False)
```

Comment:
Do we need to check for deletion?

```suggestion
            metadata = metadata_or(request.name)
            is_ready = metadata.get("ready", False)
            deleted_at = metadata.get("deleted_at", None)
            if deleted_at:
                context.abort....
```

Reply:
this is not the source, why checking the deletion?

Reply:
In case creating is taking too long (non-cow) and user decides to delete the pvc? Though I guess we'd also need to cancel the creation task itself

Reply:
In that case it will call delete volume and it will get locked

Reply:
I think our volume deletion process is buggy I will change it a bit, in terms of lock management and things like that

Reply:
At first we are taking lock
then we remove data dir and data
and after that we remove and release the lock

I will change the order a bit and the delete mark should happen with the lock

Reply:
Done.

I have done some protections